include/libpmemobj++/experimental/mpsc_queue.hpp

```diff
@@ -156,27 +195,126 @@ class mpsc_queue {
+						      w, req_size);
+
+			if (offset == -1)
+				return false;
+
+			f(range);
+
+			store_to_log(pmem::obj::string_view(data.get(), size),
+				     queue->buf + offset);
+
+			ringbuf_produce(queue->ring_buffer.get(), w);
+
+			return true;
+		}
+
+		bool
+		try_produce(pmem::obj::string_view data)
+		{
+			auto req_size = pmem::detail::align_up(
+				data.size() + sizeof(first_block::size),
+				CACHELINE_SIZE);
+			auto offset = ringbuf_acquire(queue->ring_buffer.get(),
+						      w, req_size);
+
+			if (offset == -1)
+				return false;
+
+			store_to_log(data, queue->buf + offset);
+
+			ringbuf_produce(queue->ring_buffer.get(), w);
 
-				ringbuf_produce(queue->ring_buffer.get(), w);
-				return true;
+			return true;
+		}
+
+	private:
+		void
+		store_to_log(pmem::obj::string_view data, char *log_data)
+		{
+			assert(reinterpret_cast<uintptr_t>(log_data) %
+				       CACHELINE_SIZE ==
+			       0);
+
+			first_block fblock;
+			fblock.size = 0;
+
+			/*
+			 * Depending on the size of the source data, we might
+			 * need to perform up to three separate copies:
+			 * 1. The first cacheline, 8b of metadata and 56b
+			 *	of data If there's still data to be logged:
+			 * 2. The entire remainder of data data aligned
+			 *	down to cacheline, for example, if there's 150b
+			 *	left, this step will copy only 128b. Now, we are
+			 *	left with between 0 to 63 bytes. If nonzero:
+			 * 3. Create a stack allocated cacheline-sized
+			 *	buffer, fill in the remainder of the data, and
+			 *	copy the entire cacheline.
+			 *
+			 * This is done so that we avoid a cache-miss on
+			 * misaligned writes.
+			 */
+
+			size_t ncopy =
+				(std::min)(data.size(), first_block::CAPACITY);
+			std::copy_n(data.data(), ncopy, fblock.data);
+
+			size_t remaining_size =
+				ncopy > data.size() ? 0 : data.size() - ncopy;
+
+			const char *srcof = data.data() + ncopy;
+			size_t rcopy = pmem::detail::align_down(remaining_size,
+								CACHELINE_SIZE);
+			size_t lcopy = remaining_size - rcopy;
+
+			char last_cacheline[CACHELINE_SIZE];
+			if (lcopy != 0)
+				std::copy_n(srcof + rcopy, lcopy,
+					    last_cacheline);
+
+			if (rcopy != 0) {
+				char *dest = log_data + CACHELINE_SIZE;
+
+				pmemobj_memcpy(
+					queue->pop.handle(), dest, srcof, rcopy,
+					PMEMOBJ_F_MEM_NODRAIN |
+						PMEMOBJ_F_MEM_NONTEMPORAL);
 			}
-			return false;
+
+			if (lcopy != 0) {
+				void *dest = log_data + CACHELINE_SIZE + rcopy;
+
+				pmemobj_memcpy(
+					queue->pop.handle(), dest,
+					last_cacheline, CACHELINE_SIZE,
+					PMEMOBJ_F_MEM_NODRAIN |
+						PMEMOBJ_F_MEM_NONTEMPORAL);
+			}
+
+			pmemobj_memcpy(queue->pop.handle(), log_data,
+				       reinterpret_cast<char *>(&fblock),
+				       CACHELINE_SIZE,
+				       PMEMOBJ_F_MEM_NODRAIN |
+					       PMEMOBJ_F_MEM_NONTEMPORAL);
+
+			pmemobj_drain(queue->pop.handle());
+
+			auto b = reinterpret_cast<first_block *>(log_data);
+
+			b->size = data.size();
```

Comment:
this is the fourth store, so the comment above is not accurate.
And, given that you have all the data already, you might as well just modify the `fblock` and do another copy of the entire cacheline - that will probably be faster than doing a temporal store here (since that will cause 2 memory operations).